modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerImpl.java

```diff
@@ -1611,4 +1626,45 @@ private Function<Optional<Event>, Optional<Event>> getEventUpdateFunction(Snapsh
       return Optional.of(event);
     };
   }
+
+  private Function<Optional<Event>, Optional<Event>> getEventUpdateFunctionOnlyAcl(Snapshot snapshot,
+      String orgId, User user) {
+    return (Optional<Event> eventOpt) -> {
+      MediaPackage mp = snapshot.getMediaPackage();
+      String eventId = mp.getIdentifier().toString();
+      Event event = eventOpt.orElse(new Event(eventId, orgId));
+
+      event = updateAclInEvent(event, mp, eventId);
+
+      return Optional.of(event);
+    };
+  }
+
+  private Event updateAclInEvent(Event event, MediaPackage mp, String eventId) {
+    AccessControlList acl = authorizationService.getActiveAcl(mp).getA();
+    List<ManagedAcl> acls = aclServiceFactory.serviceFor(securityService.getOrganization()).getAcls();
+
+    if (episodeIdRole) {
+      // Add custom roles to the ACL
+      // This allows users with a role of the form ROLE_EPISODE_<ID>_<ACTION> to access the event through the index
+      List<AccessControlEntry> customEntries = new ArrayList<>();
+      for (AccessControlEntry entry : acl.getEntries()) {
```

Comment:
The intention here is to also have roles for other actions. So you can for example have

```
ROLE_EPISODE_abc123_ANNOTATE,
ROLE_EPISODE_abc123_ANNOTATE_ADMIN,
```

as well. But maybe the code is flawed anyway?

Reply:
That being said,

> Why are we doing this once for every ACL entry that's already existing?

is probably a valid question anyway. With how I wrote this, I rely on all the actions I want `ROLE_EPISODE_[...]` for to be present in the ACL already. Maybe that's bad? Maybe I want to generate `ROLE_EPISODE_[...]` for all possible actions defined in the system instead? Is there a list of those somewhere?

Reply:
Not that I'm aware of, and certainly if someone were to add a new one that list would likely be incomplete.  I think depending on the incoming ACL seems reasonable, since it should contain everything that's needed.  I guess the question becomes how many is too many to process.  I hope people don't have events with hundreds of ACL entries, but it's technically possible.

I would also be tempted to either check if the generated role is already in `customEntries`, or make `customEntries` a `Set` so that we don't generate a ton of dup roles.  Past that I'm not sure how else to do what you're doing.

Reply:
Okay, the first concern was already brought up: I think this would generate duplicate roles, because we generate the new roles for each entry, and don't check whether any id roles already exists either (although the latter is probably unlikely?) This can create interesting effects for example in the Admin UI.

Second, granting read access for write access is something I feel unsure about. I know we do this in a couple of places, but for these roles, the ES index would be the only place where they are being handled this way, right? I don't know if we want that.

As Greg said, this simply won't work for any new actions that are introduced afterwards, unless an index update is triggered somehow. That's not great, but it is the way it is and as long as we document this, I think it's okay.

What I really don't like is depending on the existing ACL to set these new id roles. This will lead to inconsistencies since not every action needs to be present in every ACL, which will mean that the new roles will work for some events and not others. That's a support nightmare.

Custom actions are defined in the listprovider `acl.additional.actions.properties` so we should probably pull them from there. Read and write roles can always be generated. And we could add a comment to that config file to tell people they have to re-index the ACLs if they add a new action and use the id roles.

Also 

> I hope people don't have events with hundreds of ACL entries

Oh, some definitely do. :upside_down_face: